File: televault-desktop/packages/televault-core/src/captionCodec.js
const { createVaultEntry, dirMarker } = require('./vaultEntry');

const CAPTION_VERSION = 1;

/** @param {import('./vaultEntry').VaultEntry} entry */
function encodeCaption(entry) {
  if (entry.path.endsWith('/')) {
    const payload = { v: CAPTION_VERSION, dir: entry.path };
    if (entry.tags.length > 0) payload.tags = entry.tags;
    return JSON.stringify(payload);
  }
  return JSON.stringify({
    v: CAPTION_VERSION,
    path: entry.path,
    size: entry.size,
    sha256: entry.sha256,
    mtime: entry.mtime.toISOString(),
  });
}

/**
 * @param {number} messageId
 * @param {string} caption
 * @returns {import('./vaultEntry').VaultEntry|null}
 */
function decodeCaption(messageId, caption) {
  let m;
  try {
    const d = JSON.parse(caption);
    if (!d || typeof d !== 'object') return null;
    m = d;
  } catch {
    return null;
  }
  if (m.v !== CAPTION_VERSION) return null;

  if (typeof m.dir === 'string' && m.dir.startsWith('/') && m.dir.endsWith('/')) {
    const tagsRaw = m.tags;
    const tags = Array.isArray(tagsRaw) ? tagsRaw.filter((t) => typeof t === 'string') : [];
    return dirMarker({ messageId, path: m.dir, tags });
  }

  const path = m.path;
  if (typeof path !== 'string' || !path.startsWith('/') || path.endsWith('/')) return null;
  const mtimeRaw = m.mtime;
  const mtime = typeof mtimeRaw === 'string' ? new Date(mtimeRaw) : null;
  if (!mtime || Number.isNaN(mtime.getTime())) return null;

  const size = typeof m.size === 'number' ? m.size : 0;
  const sha256 = typeof m.sha256 === 'string' ? m.sha256 : '';

  return createVaultEntry({ messageId, path, size, sha256, mtime, tags: [] });
}

module.exports = { CAPTION_VERSION, encodeCaption, decodeCaption };

File: televault/lib/models/caption_codec.dart
import 'dart:convert';

import 'vault_entry.dart';

const int captionVersion = 1;

String encodeCaption(VaultEntry e) {
  if (e.isDir) {
    return jsonEncode({
      'v': captionVersion,
      'dir': e.path,
      if (e.tags.isNotEmpty) 'tags': e.tags,
    });
  }
  return jsonEncode({
    'v': captionVersion,
    'path': e.path,
    'size': e.size,
    'sha256': e.sha256,
    'mtime': e.mtime.toUtc().toIso8601String(),
  });
}

/// Trả null nếu caption không phải metadata hợp lệ của app (không throw).
VaultEntry? decodeCaption(int messageId, String caption) {
  Map<String, dynamic> m;
  try {
    final d = jsonDecode(caption);
    if (d is! Map<String, dynamic>) return null;
    m = d;
  } catch (_) {
    return null;
  }
  if (m['v'] != captionVersion) return null;

  final dir = m['dir'];
  if (dir is String && dir.startsWith('/') && dir.endsWith('/')) {
    final tagsRaw = m['tags'];
    return VaultEntry.dirMarker(
      messageId: messageId,
      path: dir,
      tags: tagsRaw is List ? tagsRaw.whereType<String>().toList() : const [],
    );
  }

  final path = m['path'];
  if (path is! String || !path.startsWith('/') || path.endsWith('/')) return null;
  final mtimeRaw = m['mtime'];
  final mtime = mtimeRaw is String ? DateTime.tryParse(mtimeRaw) : null;
  if (mtime == null) return null;
  final sizeRaw = m['size'];
  final shaRaw = m['sha256'];
  final tagsRaw = m['tags'];
  return VaultEntry(
    messageId: messageId,
    path: path,
    size: sizeRaw is num ? sizeRaw.toInt() : 0,
    sha256: shaRaw is String ? shaRaw : '',
    mtime: mtime.toUtc(),
    tags: tagsRaw is List ? tagsRaw.whereType<String>().toList() : const [],
  );
}
